### PR TITLE
[feat][fsdp] Add QLoRA and Gemma 4 training support

### PR DESCRIPTION
**Pull request by bvolpato**
## Summary

Add opt-in bitsandbytes 4-bit loading for FSDP policy and reference models so SkyRL can train LoRA adapters against a quantized base model in SFT and RL. Include a runnable Gemma 4 E2B single-GPU example.

## Changes

- Add `model.bitsandbytes_4bit` with NF4/FP4 and double-quantization controls.
- Load policy and reference models with BF16 compute and quantization storage.
- Keep bitsandbytes `Params4bit` off the meta-device path and avoid redundant single-rank FSDP wrapping.
- Reject FSDP-native CPU offload for unwrapped single-rank 4-bit models and direct users to manual phase offload.
- Preserve packed 4-bit payloads during multi-rank FSDP2 all-gathers by disabling FSDP parameter casting while retaining BF16 compute and FP32 gradient reduction.
- Prevent embeddings inside selected parent modules from being wrapped twice.
- Restrict `language_model_only` LoRA adapters to the language model.
- Support explicit LoRA target and exclusion lists.
- Expose SFT FlashAttention selection, validate its padding invariant, and add a Gemma-safe SDPA example.
- Add the `qlora` extra, configuration tests, wrap-policy coverage, docs, and example script.

New behavior is disabled by default.

## Repro

After accepting the Gemma license on Hugging Face:

```bash
bash examples/train/sft/run_sft_gemma4_e2b_qlora.sh
```

## Validation

- Current focused config, SFT, and FSDP tests (`212 passed`)
- Two-rank FSDP2 QLoRA byte-preservation probe on one RTX 5070 Ti: 14 quantized modules, 23 BF16 NaN storage values, zero packed-byte mismatches, finite loss and gradients, and successful optimizer step
- Full pre-commit and `uv lock --check`
- Fresh 4-bit Gemma 4 E2B load on RTX 5070 Ti: 525 quantized parameters, finite BF16 logits, shape `(1, 2, 262144)`
- Prior SkyRL Train CPU suite (`1402 passed, 5 skipped, 5 deselected`)
- Real 4-bit reference-worker load and finite forward
- Gemma 4 E2B 100-step QLoRA run: held-out loss `4.2274` to `1.1558`; finite gradients throughout
- Exported 38 MB PEFT adapter reloaded over the same 4-bit base with finite logits
- Gemma 4 E4B 100-step run: held-out loss `2.7891` to `1.1067`
- Combined with #1974, exact-model GLM-5.2 GRPO QLoRA completed rollout, backward, optimizer updates, and adapter sync with finite gradients

Single-rank and colocated RL paths were validated on a 16 GB RTX 5070 Ti. Multi-rank FSDP2 was validated with two Gloo ranks sharing that GPU; separate-GPU NCCL remains untested locally.

---

> [!NOTE]
> **Medium Risk**
> Changes core FSDP initialization and mixed-precision behavior for a new training mode; mistakes could corrupt quantized weights or break distributed training, though the feature is off by default and guarded by config validation.
> 
> **Overview**
> Adds **opt-in FSDP QLoRA**: 4-bit bitsandbytes base weights with LoRA adapters for SFT and RL policy/ref workers, controlled by `model.bitsandbytes_4bit` and the new `qlora` install extra.
> 
> **FSDP2** is taught not to cast packed `Params4bit` (avoids corrupting NF4 storage), skips meta-device init for quantized loads, shards 4-bit models in place on multi-GPU, and leaves single-GPU 4-bit models unwrapped (with a hard error if `fsdp_config.cpu_offload=true`). Embedding wrap policy no longer double-shards embeddings already inside a wrapped transformer block.
> 
> **Model loading** wires quant type/double-quant into `HFModelWrapper`, requires LoRA when 4-bit training is on, and tightens `language_model_only` LoRA to the language-model subtree. SFT gains an explicit `flash_attn` flag plus validation that padding removal needs FlashAttention; the Gemma 4 E2B example uses SDPA instead.
> 
> Also includes LoRA config list support, safer PEFT `exclude_modules` serialization, docs, tests, and `uv.lock` for bitsandbytes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4b8bde7c6126841e8541ff2d92030477c87223b0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: skyrl/backends/skyrl_train/workers/model_wrapper.py [hunk 1]
```diff
@@ -33,6 +34,24 @@
...
+    prefixes = [
+        name for name, _ in model.named_modules() if name == "language_model" or name.endswith(".language_model")
+    ]
+    if not prefixes:
+        return exclude_modules
+
+    prefix = min(prefixes, key=len)
+    outside_language_model = rf"^(?!{re.escape(prefix)}(?:\.|$)).*$"
+    if exclude_modules is None:
+        return outside_language_model
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/skyrl_train/workers/model_wrapper.py]: Updated helper to return outside-language-model pattern directly for falsey exclusions. PEFT 0.18.1 applies string exclusions with re.fullmatch, so empty alternative matched only empty module names rather than all modules; change still removes ambiguous pattern.

File: skyrl/backends/skyrl_train/distributed/fsdp_utils.py [hunk 2]
```diff
@@ -224,11 +224,18 @@ def apply_fsdp2(model, fsdp_kwargs, config: Union[FSDPConfig, DictConfig]):
...
+        if module.__class__.__name__ in fsdp_transformer_layer_cls_to_wrap:
             modules.append(module)
+            wrapped_module_names.append(name)
+
+    for name, module in model.named_modules():
+        if not isinstance(module, nn.Embedding) or model.config.tie_word_embeddings:
+            continue
+        if any(name.startswith(f"{parent_name}.") for parent_name in wrapped_module_names if parent_name):
+            continue
+        modules.append(module)
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/skyrl_train/distributed/fsdp_utils.py:240]: Collapsed module and embedding collection into one named_modules traversal, then filtered nested embeddings. Config access now safely defaults tie_word_embeddings to false.

File: skyrl/train/config/config.py [hunk 3]
```diff
@@ -1520,6 +1533,15 @@ def __post_init__(self):
...
             )
 
+        if self.policy.model.bitsandbytes_4bit.enabled:
+            assert (
+                self.strategy == "fsdp"
+            ), "`trainer.policy.model.bitsandbytes_4bit.enabled=True` is only supported with `trainer.strategy=fsdp`."
+            assert self.policy.model.lora.rank > 0, (
+                "`trainer.policy.model.bitsandbytes_4bit.enabled=True` requires "
+                "`trainer.policy.model.lora.rank > 0` for QLoRA."
+            )
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/train/config/config.py]: Replaced both QLoRA configuration assertions with ValueError checks, preserving validation under python -O.

File: skyrl/backends/skyrl_train/workers/fsdp/fsdp_worker.py [hunk 4]
```diff
@@ -147,18 +147,24 @@ def init_model(self, model_path, num_training_steps: int = None):
...
         )
 
         wrapped_model = HFModelWrapper(
             model_path,
             use_flash_attention_2=self.cfg.flash_attn,
-            bf16=self.cfg.policy.inference_only_init,
+            bf16=self.cfg.policy.inference_only_init or quantization.enabled,
+            load_in_4bit=quantization.enabled,
+            bnb_4bit_quant_type=quantization.quant_type,
+            bnb_4bit_use_double_quant=quantization.use_double_quant,
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/skyrl_train/workers/fsdp/fsdp_worker.py:173]: Fixed in fbd41bbf. Reference worker now reads its 4-bit config, passes it into FSDP strategy and model wrapper, and skips meta-device handling for Params4bit. Validated with a real Gemma 4 E2B 4-bit reference load and finite forward on an RTX 5070 Ti.

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 5]
```diff
@@ -221,6 +222,16 @@ def _fsdp_init_model(self, model, is_train=True, is_wrapped=False):
...
         module = model.model if is_wrapped else model
+
+        if self.is_4bit and self.world_size == 1:
+            return module
+
+        # Params4bit cannot be moved through the meta device. Each rank already
+        # loaded the same quantized checkpoint, so shard it in place instead.
+        if self.is_4bit:
+            apply_fsdp2(module, fsdp_kwargs, self.fsdp_config)
+            return module
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/skyrl_train/distributed/fsdp_strategy.py:246]: Fixed in 61d91e26. Reproduced on two real FSDP2 ranks before the fix: 8 of 14 NF4 modules had altered packed bytes on both ranks. Setting `MixedPrecisionPolicy.param_dtype=None` for 4-bit storage preserved all 14 modules byte-for-byte while keeping finite BF16 forward, gradients, and optimizer step. Focused config, SFT, and FSDP suites: 210 passed; pre-commit passed.

File: skyrl/backends/skyrl_train/distributed/fsdp_strategy.py [hunk 6]
```diff
@@ -221,6 +229,16 @@ def _fsdp_init_model(self, model, is_train=True, is_wrapped=False):
             "reshard_after_forward": self.fsdp_config.reshard_after_forward,
         }
         module = model.model if is_wrapped else model
+
+        if self.is_4bit and self.world_size == 1:
+            return module
```
[the remark this answers is not in the document]
Comment (bvolpato) [reply skyrl/backends/skyrl_train/distributed/fsdp_strategy.py:240]: Valid. Single-rank 4-bit models skip FSDP wrapping, so native CPU offload cannot apply. The branch now fails fast and directs users to `fsdp_config.cpu_offload=false`, which keeps SkyRL manual phase offload active. Focused FSDP/config suite: 212 passed.
